src/codec.rs

```diff
@@ -124,18 +126,20 @@ impl Decoder for LanguageServerCodec {
     }
 }
 
-fn parse_message(input: &str) -> IResult<&str, String> {
+fn parse_message(input: &[u8]) -> IResult<&[u8], &[u8]> {
     let content_len = delimited(tag("Content-Length: "), digit1, crlf);
 
     let utf8 = alt((tag("utf-8"), tag("utf8")));
     let charset = tuple((char(';'), space0, tag("charset="), utf8));
     let content_type = tuple((tag("Content-Type:"), is_not(";\r"), opt(charset), crlf));
 
     let header = terminated(terminated(content_len, opt(content_type)), crlf);
-    let length = map_res(header, |s: &str| s.parse::<usize>());
+    let length = map_res(header, |s: &[u8]| {
+        str::from_utf8(s).unwrap().parse::<usize>()
```

Comment:
I don't believe that `unwrap()` is appropriate here. This would mean that if `str::from_utf8(s)` fails, which it will when it receives an incomplete UTF-8 message, it will cause the LSP server to panic at runtime. I imagine that we might need to use `and_then()` or perhaps two separate calls to `map_res()` instead for this to work.

Reply:
Actually, `and_then()` won't work here because the error types returned by `str::from_utf8()` and `<usize as FromStr>::from_str()` differ. Instead, I imagine we might need two separate calls to `map_res()`, such as:

```
let length_raw = map_res(header, |s: &[u8]| str::from_utf8(s));
let length = map_res(length_raw, |s: &str| s.parse::<usize>());
```

Reply:
Thank you for the review. I fixed it.